### PR TITLE
Replace `CodableValue` with `AnyCodable`

### PR DESCRIPTION
**Pull request by maxep**
### What and why?

Share `AnyCodable` logic in `DatadogInternal`.

While migrating the WebView tracking to v2, I realise that we were doing such heavy transformation:
```swift
let jsonData = try JSONSerialization.data(withJSONObject: event, options: [])
let encodableEvent = try JSONDecoder().decode(AnyCodable.self, from: jsonData)
writer.write(value: encodableEvent)
```

The [Flight-School/AnyCodable](https://github.com/Flight-School/AnyCodable), which was already inspiring our `DatadogObjc.AnyEncodable`, is capable of serialising `Any` and so we could use that implementation to write `[String: Any]` directly.

Moreover, `CodableValue` is a similar type-erasure than [Flight-School/AnyCodable](https://github.com/Flight-School/AnyCodable) that can be use to decode JSON into `[String: Encodable]`.

For those 2 reasons, I would like to consider the `AnyCodable` as a solution for `[String: Any]` serialisation and deserialisation.

### How?

Replace `CodableValue` and `DatadogObjc.AnyEncodable` with implementation from [Flight-School/AnyCodable](https://github.com/Flight-School/AnyCodable).

Extensions for `ExpressibleBy*Literal` and `Hashable` have been stripped out, but the `Equatable` extensions remains as it can be useful for data comparison on the message bus.

#### Model Generator

The generator now uses the shared `AnyCodable` and `DynamicCodingKey`.

### Review checklist
- [ ] Feature or bugfix MUST have appropriate tests (unit, integration)
- [ ] Make sure each commit and the PR mention the Issue number or JIRA reference
- [ ] Add CHANGELOG entry for user facing changes

### Custom CI job configuration (optional)
- [x] Run unit tests
- [x] Run integration tests
- [ ] Run smoke tests

### CODE REVIEW

File: Sources/Datadog/DatadogInternal/Codable/AnyEncodable.swift [hunk 1]
```diff
@@ -0,0 +1,210 @@
...
+            try container.encode(string)
+        case let array as [Any?]:
+            try container.encode(array.map { AnyEncodable($0) })
+        case let dictionary as [String: Any?]:
+            try container.encode(dictionary.mapValues { AnyEncodable($0) })
+        case let encodable as Encodable:
+            try encodable.encode(to: encoder)
+        default:
+            let context = EncodingError.Context(codingPath: container.codingPath, debugDescription: "AnyEncodable value cannot be encoded")
+            throw EncodingError.invalidValue(value, context)
```
Comment (maxep) [inline Sources/Datadog/DatadogInternal/Codable/AnyEncodable.swift:127]: We might want to keep flexibility here and not `throw` encoding failure, WDYT?

**Comment by maxep** [comment]
This is ready for review, it works well for our different use-cases 👍  We can now encode dictionary coming from WebView tracking directly. The models have been updated to use the interface shared in `DatadogInternal`.

Some changes in `AnyEncodable` were necessary to be usable in both Swift and Objc: In both cases, a value is first cast to a `NSNumber` so it can infer the right type using the `objCType`. The reason being that the webview bridge returns a JSON as objc dictionary with `NSNumber`, in Swift a `NSNumber` with value 1 or 0 can be cast to `Bool` which would be encoded as `true` or `false` and lead to a type mismatch:

```swift
let value: Any = NSNumber(value: 1)
switch value {
case let bool as Bool:
    encode(bool) // will encode 'true'
case let nsnumber as NSNumber:
    encode(nsnumber) // won't reach :(
default:
    break
}
```

File: Sources/Datadog/FeaturesIntegration/WebView/WebRUMEventConsumer.swift [hunk 2]
```diff
@@ -40,7 +40,7 @@ internal class DefaultWebRUMEventConsumer: WebEventConsumer {
         let mappedEvent = map(event: event, with: rumContext)
 
         let jsonData = try JSONSerialization.data(withJSONObject: mappedEvent, options: [])
-        let encodableEvent = try jsonDecoder.decode(CodableValue.self, from: jsonData)
+        let encodableEvent = try jsonDecoder.decode(AnyCodable.self, from: jsonData)
```
[the remark this answers is not in the document]
Comment (maxep) [reply Sources/Datadog/FeaturesIntegration/WebView/WebRUMEventConsumer.swift:43]: Yes exactly 👍

File: tools/rum-models-generator/Tests/CodeGenerationTests/Print/ObjcInteropPrinterTests.swift [hunk 3]
```diff
@@ -1037,7 +1037,7 @@ final class ObjcInteropPrinterTests: XCTestCase {
 
         @objc
         public class DDFooBar: NSObject {
-            internal let swiftModel: Foo.Bar
+            internal var swiftModel: Foo.Bar
```
[the remark this answers is not in the document]
Comment (maxep) [reply tools/rum-models-generator/Tests/CodeGenerationTests/Print/ObjcInteropPrinterTests.swift:1040]: Yes.. It's not being tested in CI, which is ok imo, or maybe we could have a conditional workflow!

[the remark this answers is not in the document]
Comment (ncreated) [reply tools/rum-models-generator/Tests/CodeGenerationTests/Print/ObjcInteropPrinterTests.swift:1040]: We have the [`run_tools_tests`](https://github.com/DataDog/dd-sdk-ios/blob/15b5accc7b9bfa65228b5c69205a055dd46c2ec7/bitrise.yml#L436) lane which looks perfect candidate for it and [can be easily conditioned](https://github.com/DataDog/dd-sdk-ios/blob/15b5accc7b9bfa65228b5c69205a055dd46c2ec7/tools/ci/choose_workflows.py#L85-L97)]. How about creating separate task to tackle it?

File: Sources/Datadog/DatadogInternal/Codable/AnyEncodable.swift (hunk 1 above)
Comment (maxep) [reply Sources/Datadog/DatadogInternal/Codable/AnyEncodable.swift:127]: My bad, I thought we were not throwing an error in our previous `Objc.AnyEncodable`, but it did! All good then, I will add the type in the description 👍

File: tools/rum-models-generator/Tests/CodeGenerationTests/Print/ObjcInteropPrinterTests.swift (hunk 3 above)
[the remark this answers is not in the document]
Comment (maxep) [reply tools/rum-models-generator/Tests/CodeGenerationTests/Print/ObjcInteropPrinterTests.swift:1040]: @ncreated I'm adding this to CI in this PR: https://github.com/DataDog/dd-sdk-ios/pull/1081/commits/362fa40eb484788ef6171866525e17499e9467e0

**Pull request merged**